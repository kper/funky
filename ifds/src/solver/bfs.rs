use crate::icfg::graph::{Fact, Graph};
use crate::icfg::state::State;
use crate::solver::*;
use std::collections::VecDeque;

use log::debug;

pub struct Bfs;

impl GraphReachability for Bfs {
    fn all_sinks(&mut self, graph: &mut Graph, state: &State, req: &Request) -> Vec<Taint> {
        let mut queue: VecDeque<&Fact> = VecDeque::new();
        let mut seen = Vec::new();

        if let Some(start) = state
            .get_facts_at(&req.function, req.pc)
            .unwrap()
            .find(|x| &x.belongs_to_var == req.variable.as_ref().unwrap())
        {
            debug!("Adding start {:#?}", start);
            queue.push_back(start);
        } else {
            return Vec::new();
        }

        while let Some(node) = queue.pop_front() {
            debug!("Popping node {:?}", node);
            seen.push(node);
            for child in graph
                .edges
                .iter()
                .filter(|x| x.get_from() == node)
                .map(|x| x.to())
            {
                //debug!("Adding child {:#?}", graph.query_by_fact_id(child));

                if !seen.contains(&child) {
                    debug!("queue child {:?}", child);
                    queue.push_back(child);
                }
            }
        }

        seen.into_iter()
            .map(|x| Taint {
                variable: x.belongs_to_var.clone(),
                pc: x.next_pc,
                function: x.function.clone(),
            })
            .collect()
    }
}
