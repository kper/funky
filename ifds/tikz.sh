#!/bin/bash

rm /tmp/graph.tex
rm /tmp/graph.pdf
tail -n +6 $1 > /tmp/graph.tex
xelatex /tmp/graph.tex && evince graph.pdf
