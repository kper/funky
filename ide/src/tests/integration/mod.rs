use crate::ir::wasm_ast::IR;
use funky::engine::module::ModuleInstance;
use funky::engine::*;
use insta::assert_snapshot;
use validation::validate;
use wasm_parser::{parse, read_wasm};

use crate::grammar::*;
use crate::icfg::convert2::ConvertSummary;
use crate::icfg::tikz::render_to;
use crate::solver::Request;

macro_rules! wasm {
    ($input:expr) => {{
        // Read it
        let file = read_wasm!($input);
        let module = parse(file).expect("Parsing failed");
        assert!(validate(&module).is_ok());

        let imports = Vec::new();

        let instance = ModuleInstance::new(&module);
        let engine = Engine::new(
            instance,
            &module,
            Box::new(funky::debugger::RelativeProgramCounter::default()),
            &imports,
        )
        .unwrap();

        let mut ir = IR::new();

        ir.visit(&engine).unwrap();

        ir
    }};
}

macro_rules! run {
    ($name:expr, $req:expr, $fs:expr) => {
        let ir = wasm!($fs);

        let ir_code = ir.buffer();

        let mut convert = ConvertSummary::new();

        let prog = ProgramParser::new().parse(&ir_code).unwrap();

        let graph = convert.visit(&prog, &$req).unwrap();

        let output = render_to(&graph);

        assert_snapshot!(format!("{}", $name), output);
    };
}

#[test]
fn test_add() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("add", req, "../tests/add.wasm");
}

#[test]
fn test_fib() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("fib", req, "../tests/fib.wasm");
}

#[test]
fn test_fac() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("fac", req, "../tests/fac.wasm");
}

#[test]
fn test_br_table() {
    let req = Request {
        function: "0".to_string(),
        variable: None,
        pc: 0,
    };

    run!("br_table", req, "../tests/br_table.wasm");
}
