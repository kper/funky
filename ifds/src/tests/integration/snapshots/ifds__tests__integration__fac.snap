---
source: src/tests/integration/mod.rs
expression: output

---
   \documentclass{standalone}
            \usepackage{pgf, tikz}
            \usetikzlibrary{arrows, automata}
            \begin{document}
                \begin{tikzpicture}
                    [node distance=3cm]
                    \tikzstyle{every state}=[
                        draw = black,
                        thick,
                        fill = white,
                        minimum size = 4mm
                    ]

                    \node[circle,fill,inner sep=1pt,label=left:$\%10$] (156) at (11, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$\%11$] (155) at (12, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (154) at (11, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$\%11$] (153) at (12, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (152) at (11, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (151) at (11, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (150) at (11, 20) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (149) at (11, 19) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (148) at (11, 18) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (147) at (11, 17) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (146) at (11, 16) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (145) at (10, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$\%10$] (144) at (11, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$\%9$] (142) at (10, 14) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (140) at (2, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (139) at (4, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (138) at (2, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (137) at (4, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (136) at (2, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (135) at (4, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (134) at (2, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (133) at (4, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (132) at (2, 20) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (131) at (4, 20) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (130) at (2, 19) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (129) at (4, 19) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (128) at (2, 18) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (127) at (4, 18) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (126) at (2, 17) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (125) at (4, 17) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (124) at (2, 16) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (123) at (4, 16) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (122) at (2, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (121) at (4, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (120) at (2, 14) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (119) at (2, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (118) at (4, 14) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (117) at (4, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (116) at (2, 13) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (115) at (2, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (114) at (4, 13) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (113) at (4, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (112) at (2, 12) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (111) at (2, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (110) at (4, 12) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (109) at (4, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (108) at (2, 11) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (107) at (2, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (106) at (4, 11) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (105) at (4, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (104) at (2, 10) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (103) at (2, 7) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (102) at (4, 10) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (101) at (4, 7) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (100) at (2, 9) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (99) at (2, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (98) at (4, 9) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (97) at (4, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (96) at (2, 8) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (95) at (2, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (94) at (4, 8) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (93) at (4, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (92) at (2, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$\%3$] (91) at (4, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (90) at (2, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (89) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (87) at (0, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (86) at (0, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (85) at (0, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (84) at (0, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (83) at (0, 20) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (82) at (0, 19) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (81) at (0, 18) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (80) at (0, 17) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (79) at (0, 16) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (78) at (0, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (77) at (0, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (76) at (0, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (75) at (0, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (74) at (0, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (73) at (0, 20) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (72) at (0, 19) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (71) at (0, 18) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (70) at (0, 17) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (69) at (0, 16) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (68) at (0, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (67) at (0, 14) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (66) at (0, 13) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (65) at (0, 12) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (64) at (0, 11) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (63) at (0, 10) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (62) at (0, 9) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (61) at (0, 8) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (60) at (0, 7) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (59) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (58) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (57) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (56) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (55) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (54) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (53) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (52) at (0, 13) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (51) at (0, 11) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (50) at (0, 10) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (49) at (0, 9) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (48) at (0, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (47) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (46) at (0, 8) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (45) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (44) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (43) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (42) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$\%1$] (41) at (2, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (40) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (39) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (37) at (0, 24) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (36) at (0, 23) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (35) at (0, 22) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (34) at (0, 21) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (33) at (0, 20) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (32) at (0, 19) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (31) at (0, 18) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (30) at (0, 17) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (29) at (0, 16) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (28) at (0, 15) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (27) at (0, 14) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (26) at (0, 13) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (25) at (0, 12) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (24) at (0, 11) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (23) at (0, 10) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (22) at (0, 9) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (21) at (0, 8) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (20) at (0, 7) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (19) at (0, 6) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (18) at (0, 5) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (17) at (0, 4) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (16) at (0, 3) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (15) at (0, 2) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (14) at (0, 1) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (13) at (0, 0) { };
\node[circle,fill,inner sep=1pt,label=left:$taut$] (0) at (0, 0) { };
\node[font=\tiny] (note_0) at (-1.5, 0.5) {Block(0)};
\node[font=\tiny] (note_1) at (-1.5, 1.5) {Assign(1, 0)};
\node[font=\tiny] (note_2) at (-1.5, 2.5) {Const(2, 0.0)};
\node[font=\tiny] (note_3) at (-1.5, 3.5) {BinOp(3, 2, 1)};
\node[font=\tiny] (note_4) at (-1.5, 4.5) {Conditional(3, [1, 2])};
\node[font=\tiny] (note_5) at (-1.5, 5.5) {Block(1)};
\node[font=\tiny] (note_6) at (-1.5, 6.5) {Const(4, 1.0)};
\node[font=\tiny] (note_7) at (-1.5, 7.5) {Jump(3)};
\node[font=\tiny] (note_8) at (-1.5, 8.5) {Block(2)};
\node[font=\tiny] (note_9) at (-1.5, 9.5) {Assign(5, 0)};
\node[font=\tiny] (note_10) at (-1.5, 10.5) {Assign(6, 0)};
\node[font=\tiny] (note_11) at (-1.5, 11.5) {Const(7, 1.0)};
\node[font=\tiny] (note_12) at (-1.5, 12.5) {BinOp(8, 7, 6)};
\node[font=\tiny] (note_13) at (-1.5, 13.5) {Call(0, [8], [9])};
\node[font=\tiny] (note_14) at (-1.5, 14.5) {BinOp(10, 9, 8)};
\node[font=\tiny] (note_15) at (-1.5, 15.5) {Kill(9)};
\node[font=\tiny] (note_16) at (-1.5, 16.5) {Kill(8)};
\node[font=\tiny] (note_17) at (-1.5, 17.5) {Kill(7)};
\node[font=\tiny] (note_18) at (-1.5, 18.5) {Kill(6)};
\node[font=\tiny] (note_19) at (-1.5, 19.5) {Kill(5)};
\node[font=\tiny] (note_20) at (-1.5, 20.5) {Jump(3)};
\node[font=\tiny] (note_21) at (-1.5, 21.5) {Block(3)};
\node[font=\tiny] (note_22) at (-1.5, 22.5) {Phi(11, 4, 10)};
\node[font=\tiny] (note_23) at (-1.5, 23.5) {Return([11])};
\node[font=\tiny] (note_24) at (-1.5, 0.5) {Block(0)};
\node[font=\tiny] (note_25) at (-1.5, 1.5) {Assign(1, 0)};
\node[font=\tiny] (note_26) at (-1.5, 2.5) {Const(2, 0.0)};
\node[font=\tiny] (note_27) at (-1.5, 3.5) {BinOp(3, 2, 1)};
\node[font=\tiny] (note_28) at (-1.5, 4.5) {Conditional(3, [1, 2])};
\node[font=\tiny] (note_29) at (-1.5, 5.5) {Block(1)};
\node[font=\tiny] (note_30) at (-1.5, 6.5) {Const(4, 1.0)};
\node[font=\tiny] (note_31) at (-1.5, 7.5) {Jump(3)};
\node[font=\tiny] (note_32) at (-1.5, 8.5) {Block(2)};
\node[font=\tiny] (note_33) at (-1.5, 9.5) {Assign(5, 0)};
\node[font=\tiny] (note_34) at (-1.5, 10.5) {Assign(6, 0)};
\node[font=\tiny] (note_35) at (-1.5, 11.5) {Const(7, 1.0)};
\node[font=\tiny] (note_36) at (-1.5, 12.5) {BinOp(8, 7, 6)};
\node[font=\tiny] (note_37) at (-1.5, 13.5) {Call(0, [8], [9])};
\node[font=\tiny] (note_38) at (-1.5, 14.5) {BinOp(10, 9, 8)};
\node[font=\tiny] (note_39) at (-1.5, 15.5) {Kill(9)};
\node[font=\tiny] (note_40) at (-1.5, 16.5) {Kill(8)};
\node[font=\tiny] (note_41) at (-1.5, 17.5) {Kill(7)};
\node[font=\tiny] (note_42) at (-1.5, 18.5) {Kill(6)};
\node[font=\tiny] (note_43) at (-1.5, 19.5) {Kill(5)};
\node[font=\tiny] (note_44) at (-1.5, 20.5) {Jump(3)};
\node[font=\tiny] (note_45) at (-1.5, 21.5) {Block(3)};
\node[font=\tiny] (note_46) at (-1.5, 22.5) {Phi(11, 4, 10)};
\node[font=\tiny] (note_47) at (-1.5, 23.5) {Return([11])};
		\path[->, blue] (0) [loop right] edge  node { } (0);
		\path[->, blue] (0) [bend right] edge  node { } (14);
		\path[->, blue] (0) [bend right] edge  node { } (15);
		\path[->, blue] (0) [bend right] edge  node { } (16);
		\path[->, blue] (0) [bend right] edge  node { } (17);
		\path[->, blue] (0) [bend right] edge  node { } (18);
		\path[->, blue] (0) [bend right] edge  node { } (19);
		\path[->, blue] (0) [bend right] edge  node { } (20);
		\path[->, blue] (0) [bend right] edge  node { } (21);
		\path[->, blue] (0) [bend right] edge  node { } (22);
		\path[->, blue] (0) [bend right] edge  node { } (23);
		\path[->, blue] (0) [bend right] edge  node { } (24);
		\path[->, blue] (0) [bend right] edge  node { } (25);
		\path[->, blue] (0) [bend right] edge  node { } (26);
		\path[->, blue] (0) [bend right] edge  node { } (27);
		\path[->, blue] (0) [bend right] edge  node { } (28);
		\path[->, blue] (0) [bend right] edge  node { } (29);
		\path[->, blue] (0) [bend right] edge  node { } (30);
		\path[->, blue] (0) [bend right] edge  node { } (31);
		\path[->, blue] (0) [bend right] edge  node { } (32);
		\path[->, blue] (0) [bend right] edge  node { } (33);
		\path[->, blue] (0) [bend right] edge  node { } (34);
		\path[->, blue] (0) [bend right] edge  node { } (35);
		\path[->, blue] (0) [bend right] edge  node { } (36);
		\path[->, blue] (0) [bend right] edge  node { } (37);
		\path[->, blue] (39) [bend right] edge  node { } (40);
		\path[->, blue] (39) [bend right] edge  node { } (41);
		\path[->, blue] (39) [bend right] edge  node { } (89);
		\path[->, blue] (39) [bend right] edge  node { } (90);
		\path[->, blue] (39) [bend right] edge  node { } (91);
		\path[->, blue] (39) [bend right] edge  node { } (92);
		\path[->, blue] (39) [bend right] edge  node { } (93);
		\path[->, blue] (39) [bend right] edge  node { } (94);
		\path[->, blue] (39) [bend right] edge  node { } (95);
		\path[->, blue] (39) [bend right] edge  node { } (96);
		\path[->, blue] (39) [bend right] edge  node { } (97);
		\path[->, blue] (39) [bend right] edge  node { } (98);
		\path[->, blue] (39) [bend right] edge  node { } (99);
		\path[->, blue] (39) [bend right] edge  node { } (100);
		\path[->, blue] (39) [bend right] edge  node { } (101);
		\path[->, blue] (39) [bend right] edge  node { } (102);
		\path[->, blue] (39) [bend right] edge  node { } (103);
		\path[->, blue] (39) [bend right] edge  node { } (104);
		\path[->, blue] (39) [bend right] edge  node { } (105);
		\path[->, blue] (39) [bend right] edge  node { } (106);
		\path[->, blue] (39) [bend right] edge  node { } (107);
		\path[->, blue] (39) [bend right] edge  node { } (108);
		\path[->, blue] (39) [bend right] edge  node { } (109);
		\path[->, blue] (39) [bend right] edge  node { } (110);
		\path[->, blue] (39) [bend right] edge  node { } (111);
		\path[->, blue] (39) [bend right] edge  node { } (112);
		\path[->, blue] (39) [bend right] edge  node { } (113);
		\path[->, blue] (39) [bend right] edge  node { } (114);
		\path[->, blue] (39) [bend right] edge  node { } (115);
		\path[->, blue] (39) [bend right] edge  node { } (116);
		\path[->, blue] (39) [bend right] edge  node { } (117);
		\path[->, blue] (0) [bend right] edge  node { } (118);
		\path[->, blue] (39) [bend right] edge  node { } (119);
		\path[->, blue] (0) [bend right] edge  node { } (120);
		\path[->, blue] (0) [bend right] edge  node { } (121);
		\path[->, blue] (0) [bend right] edge  node { } (122);
		\path[->, blue] (0) [bend right] edge  node { } (123);
		\path[->, blue] (0) [bend right] edge  node { } (124);
		\path[->, blue] (0) [bend right] edge  node { } (125);
		\path[->, blue] (0) [bend right] edge  node { } (126);
		\path[->, blue] (0) [bend right] edge  node { } (127);
		\path[->, blue] (0) [bend right] edge  node { } (128);
		\path[->, blue] (0) [bend right] edge  node { } (129);
		\path[->, blue] (0) [bend right] edge  node { } (130);
		\path[->, blue] (0) [bend right] edge  node { } (131);
		\path[->, blue] (0) [bend right] edge  node { } (132);
		\path[->, blue] (0) [bend right] edge  node { } (133);
		\path[->, blue] (0) [bend right] edge  node { } (134);
		\path[->, blue] (0) [bend right] edge  node { } (135);
		\path[->, blue] (0) [bend right] edge  node { } (136);
		\path[->, blue] (0) [bend right] edge  node { } (137);
		\path[->, blue] (0) [bend right] edge  node { } (138);
		\path[->, blue] (0) [bend right] edge  node { } (139);
		\path[->, blue] (0) [bend right] edge  node { } (140);
		\path[->, blue] (0) [bend right] edge  node { } (142);
		\path[->, blue] (0) [bend right] edge  node { } (144);
		\path[->, blue] (0) [bend right] edge  node { } (145);
		\path[->, blue] (0) [bend right] edge  node { } (146);
		\path[->, blue] (0) [bend right] edge  node { } (147);
		\path[->, blue] (0) [bend right] edge  node { } (148);
		\path[->, blue] (0) [bend right] edge  node { } (149);
		\path[->, blue] (0) [bend right] edge  node { } (150);
		\path[->, blue] (0) [bend right] edge  node { } (151);
		\path[->, blue] (0) [bend right] edge  node { } (152);
		\path[->, blue] (0) [bend right] edge  node { } (153);
		\path[->, blue] (0) [bend right] edge  node { } (154);
		\path[->, blue] (0) [bend right] edge  node { } (155);
		\path[->, blue] (0) [bend right] edge  node { } (156);
		\path[->] (0) edge (13);
		\path[->] (13) edge (14);
		\path[->] (14) edge (15);
		\path[->] (15) edge (16);
		\path[->] (16) edge (17);
		\path[->] (17) edge (18);
		\path[->] (18) edge (19);
		\path[->] (19) edge (20);
		\path[->] (20) edge (21);
		\path[->] (21) edge (22);
		\path[->] (22) edge (23);
		\path[->] (23) edge (24);
		\path[->] (24) edge (25);
		\path[->] (25) edge (26);
		\path[->] (26) edge (27);
		\path[->] (27) edge (28);
		\path[->] (28) edge (29);
		\path[->] (29) edge (30);
		\path[->] (30) edge (31);
		\path[->] (31) edge (32);
		\path[->] (32) edge (33);
		\path[->] (33) edge (34);
		\path[->] (34) edge (35);
		\path[->] (35) edge (36);
		\path[->] (36) edge (37);
		\path[->] (0) edge (39);
		\path[->] (39) edge (41);
		\path[->] (0) edge (42);
		\path[->] (14) edge (43);
		\path[->] (16) edge (44);
		\path[->, bend right] (17) edge (45);
		\path[->, bend right] (17) edge (46);
		\path[->] (18) edge (47);
		\path[->, bend right] (20) edge (48);
		\path[->] (21) edge (49);
		\path[->] (22) edge (50);
		\path[->] (23) edge (51);
		\path[->] (25) edge (52);
		\path[->] (0) edge (53);
		\path[->] (53) edge (54);
		\path[->] (54) edge (55);
		\path[->] (55) edge (56);
		\path[->] (56) edge (57);
		\path[->] (57) edge (58);
		\path[->] (58) edge (59);
		\path[->] (59) edge (60);
		\path[->] (60) edge (61);
		\path[->] (61) edge (62);
		\path[->] (62) edge (63);
		\path[->] (63) edge (64);
		\path[->] (64) edge (65);
		\path[->] (65) edge (66);
		\path[->] (66) edge (67);
		\path[->] (67) edge (68);
		\path[->] (68) edge (69);
		\path[->] (69) edge (70);
		\path[->] (70) edge (71);
		\path[->] (71) edge (72);
		\path[->] (72) edge (73);
		\path[->] (73) edge (74);
		\path[->] (74) edge (75);
		\path[->] (75) edge (76);
		\path[->] (76) edge (77);
		\path[->] (27) edge (78);
		\path[->] (28) edge (79);
		\path[->] (29) edge (80);
		\path[->] (30) edge (81);
		\path[->] (31) edge (82);
		\path[->] (32) edge (83);
		\path[->, bend right] (33) edge (84);
		\path[->] (34) edge (85);
		\path[->] (35) edge (86);
		\path[->] (36) edge (87);
		\path[->] (14) edge (89);
		\path[->] (41) edge (90);
		\path[->] (90) edge (91);
		\path[->] (90) edge (92);
		\path[->, bend right] (91) edge (93);
		\path[->, bend right] (91) edge (94);
		\path[->, bend right] (92) edge (95);
		\path[->, bend right] (92) edge (96);
		\path[->] (93) edge (97);
		\path[->] (94) edge (98);
		\path[->] (95) edge (99);
		\path[->] (96) edge (100);
		\path[->] (97) edge (101);
		\path[->] (98) edge (102);
		\path[->] (99) edge (103);
		\path[->] (100) edge (104);
		\path[->, bend right] (101) edge (105);
		\path[->] (102) edge (106);
		\path[->, bend right] (103) edge (107);
		\path[->] (104) edge (108);
		\path[->] (105) edge (109);
		\path[->] (106) edge (110);
		\path[->] (107) edge (111);
		\path[->] (108) edge (112);
		\path[->] (109) edge (113);
		\path[->] (110) edge (114);
		\path[->] (111) edge (115);
		\path[->] (112) edge (116);
		\path[->] (113) edge (117);
		\path[->] (114) edge node { } (118);
		\path[->] (115) edge (119);
		\path[->] (116) edge node { } (120);
		\path[->] (118) edge (121);
		\path[->] (120) edge (122);
		\path[->] (121) edge (123);
		\path[->] (122) edge (124);
		\path[->] (123) edge (125);
		\path[->] (124) edge (126);
		\path[->] (125) edge (127);
		\path[->] (126) edge (128);
		\path[->] (127) edge (129);
		\path[->] (128) edge (130);
		\path[->] (129) edge (131);
		\path[->] (130) edge (132);
		\path[->, bend right] (131) edge (133);
		\path[->, bend right] (132) edge (134);
		\path[->] (105) edge (135);
		\path[->] (107) edge (136);
		\path[->] (109) edge (137);
		\path[->] (111) edge (138);
		\path[->] (113) edge (139);
		\path[->] (115) edge (140);
		\path[->] (142) edge (144);
		\path[->] (142) edge (145);
		\path[->] (144) edge (146);
		\path[->] (146) edge (147);
		\path[->] (147) edge (148);
		\path[->] (148) edge (149);
		\path[->] (149) edge (150);
		\path[->, bend right] (150) edge (151);
		\path[->] (151) edge (152);
		\path[->] (152) edge (153);
		\path[->] (152) edge (154);
		\path[->] (153) edge (155);
		\path[->] (154) edge (156);


                \end{tikzpicture}
            \end{document}
